lib/handler/connect.c

```diff
@@ -145,13 +163,45 @@ static void record_error(struct st_connect_generator_t *self, const char *error_
             parts[nparts++] = h2o_iovec_init(H2O_STRLIT("; details="));
             parts[nparts++] = h2o_encode_sf_string(pool, details, SIZE_MAX);
         }
+        if (dest_addr_str.base != NULL) {
+            parts[nparts++] = h2o_iovec_init(H2O_STRLIT("; next-hop="));
+            parts[nparts++] = dest_addr_str;
+        }
         assert(nparts <= sizeof(parts) / sizeof(parts[0]));
         h2o_iovec_t hval = h2o_concat_list(pool, parts, nparts);
 
         h2o_add_header_by_str(pool, &self->src_req->res.headers, H2O_STRLIT("proxy-status"), 0, NULL, hval.base, hval.len);
     }
 }
 
+static void record_connect_success(struct st_connect_generator_t *self)
+{
+    if (self->server_addresses.trying == NULL)
+        return;
+    const struct st_server_address_t *addr = self->server_addresses.trying;
+
+    H2O_PROBE_REQUEST(CONNECT_SUCCESS, self->src_req, addr->sa);
+
+    if (!self->handler->config.connect_proxy_status_enabled)
+        return;
+
+    char addr_str[NI_MAXHOST];
+    size_t addr_str_len = h2o_socket_getnumerichost(addr->sa, addr->salen, addr_str);
+    if (addr_str_len == SIZE_MAX)
+        return;
+
+    h2o_iovec_t parts[] = {
+        get_proxy_status_identity(self),
+        h2o_iovec_init(H2O_STRLIT("; next-hop=")),
+        h2o_iovec_init(addr_str, addr_str_len),
+    };
+    const size_t nparts = sizeof(parts) / sizeof(parts[0]);
+    h2o_iovec_t hval = h2o_concat_list(&self->src_req->pool, parts, nparts);
```

Comment:
```suggestion
    h2o_iovec_t hval = h2o_concat(&self->src_req->pool, get_proxy_status_identity(self), h2o_iovec_init(H2O_STRLIT("; next-hop=")), h2o_iovec_init(addr_str, addr_str_len));
```

Contrary to the logic we have in `record_error`, the list being supplied here is always the same, therefore we can use `h2o_concat`.

As an alternative, I think there is a possibility of creating a unified function dedicated to building the header field, regardless of the result being success or failure. Doing so would reduce code duplication. But I think that's a matter of taste, I'm fine either ways.

Reply:
Yes, that is simpler, thank you.

I will give the unified function a try, it should work, because the success response simply does not have the `error`, `rcode` and `details` parameteers.  Thank you for the suggestion!

Reply:
Done in 332703ce5.

The call to `h2o_socket_getnumerichost` is outside of `add_proxy_status_header` because that allows the `h2o_req_log_error` call in `record_error` to log the IP address without doing the conversion twice.

What do you think?